### ISSUE TITLE
Geometry.equals returns false for identical invalid geometries

### ISSUE THREAD

**Issue by dbaston** [locationtech/jts#327]
As an example: `POLYGON ((40 170, 275 204, 120 410, 200 350, 40 170))`

Either "true" or an exception seem like preferable outcomes.

**Comment by dr-jts** [comment]
In general JTS follows the principle that the result of applying operations to invalid geometries is undefined (unless a meaningful result can be computed with no extra overhead).  The alternative is to impose a significant cost on all calls to the operation (usually, the cost of a validity check).  

If specific behaviour is needed on invalid inputs, then it's easy to build this logic using the JTS operations as building blocks.  

Possibly a good extension to JTS would be to allow geometries to carry an invalid flag, which would allow fast checking of invalidity (although with the cost of an initial invalid calculation).

Or perhaps there could be a global flag which would enable validity checking for all operations.  But that seems like a lot of work for fairly minimal payoff.

**Comment by dbaston** [comment]
I understand the general principle for invalid geometries, but isn't equality a case where a meaningful result can be computed with no extra overhead? (ie, only check for topological equality if there is not structural equality). This particular situation doesn't affect me at all, I just reported it based on a post to the PostGIS listserv.

**Comment by dr-jts** [comment]
The definition of `Geometry.equals` is to compute topological equality (as per the SFS specification).   And in any case, even testing for structural equality imposes overhead.  

There are other methods to test structural equality (e.g. `Geometry.equalsExact`).  So the same logic applies - it's easy to create simple code to do both checks if that is what is desired.

Equality is a surprisingly nebulous term when it comes to geometry.  So it's important to define exactly what is provided, and to know what you want to compute.

I'll have a look at the PostGIS post.   Perhaps there's a need there that could motivate design enhancements.

**Comment by dbaston** [comment]
The PostGIS thread is here: https://lists.osgeo.org/pipermail/postgis-users/2018-October/042907.html

Recent versions of PostGIS check structural equality before topological equality as an optimization.

The assumption is that if the inputs are topologically equal, in typical usage they will very often be structurally equal, and by checking structural equality we can avoid an expensive topological check. And if they are not structurally equal, our structural equality test will bail almost immediately (on the geometry type, or the first vertex) so it's not much of a performance penalty at all.

**Comment by dr-jts** [comment]
Good point.  Seems like a worthwhile optimization (especially in the PostGIS context, where users need this to "just work").

I'll mark this as a potential enhancement to JTS.

**Comment by dr-jts** [comment]
@dbaston Perhaps the real question is what is the use case for checking topological equality at all?  It's an expensive check, and I have rarely seen a real use case for it. 

In which case, using `equalsExact `is more precise, and will avoid the (large) overhead of attempting to compute topological equality.

Note that I"m not suggesting dropping the topological equals. It's part of the OGC spec, and it completes the set of spatial predicates.

**Comment by dbaston** [comment]
@dr-jts Another use case is finding duplicated polygons where the rings may have different origin points, or the holes may be in different orders (I run into this in messy parcel data, for example). Also useful in ignoring differences between single geometries and single-element multigeometries.

I agree that structural equality is generally more useful and probably what most people think they're getting with `ST_Equals` or `Geometry::equals`.

**Comment by dr-jts** [comment]
Good to have those use cases.  And as mentioned, those kinds of structural tests can be done much more efficiently than full topological equality.

Actually I have a theory that for valid polygons topological equality is equivalent to structural equality, with the addition of ignoring vertices which are collinear (i.e. lie exactly on the line between their neighbouring vertices).  So this points to a more efficient implementation of equality, which corresponds more directly to structural equality.  This kind of algorithm should also handle invalid polygons reasonably well.

LineString equality is a bit more complex, but can probably be handled in a similar fashion.
